aea/helpers/dialogue/base.py

```diff
@@ -412,90 +395,145 @@ def last_message(self) -> Optional[Message]:
-        return is_extendable
+            raise InvalidDialogueMessage(
+                "The message {} does not belong to this dialogue."
+                "The dialogue reference of the message is {}, while the dialogue reference of the dialogue is {}".format(
+                    message.message_id,
+                    message.dialogue_reference,
+                    self.dialogue_label.dialogue_reference,
+                )
+            )
 
-    def ensure_counterparty(self, message: Message) -> None:
-        """
-        Ensure the counterparty is set (set if not) correctly.
+        is_valid_result, is_valid_message = self.is_valid_next_message(message)
 
-        :param message: a message
-        :return: None
-        """
-        counterparty = None  # type: Optional[str]
-        try:
-            counterparty = message.counterparty
-        except AssertionError:
-            # assume message belongs to dialogue
-            message.counterparty = self.dialogue_label.dialogue_opponent_addr
+        if not is_valid_result:
+            raise InvalidDialogueMessage(
+                "Message {} is invalid with respect to this dialogue. Error: ".format(
+                    message.message_id
+                )
+                + is_valid_message
+            )
 
-        if counterparty is not None:
-            assert (
-                message.counterparty == self.dialogue_label.dialogue_opponent_addr
-            ), "The counterparty specified in the message is different from the opponent in this dialogue."
+        if self.is_message_by_self(message):
+            self._outgoing_messages.extend([message])
+        else:
+            self._incoming_messages.extend([message])
 
     def is_belonging_to_dialogue(self, message: Message) -> bool:
         """
         Check if the message is belonging to the dialogue.
 
         :param message: the message
-        :return: Ture if message is part of the dialogue, False otherwise
+        :return: True if message is part of the dialogue, False otherwise
         """
         if self.is_self_initiated:
             self_initiated_dialogue_label = DialogueLabel(
-                (message.dialogue_reference[0], Dialogue.OPPONENT_STARTER_REFERENCE),
-                message.counterparty,
+                (
+                    message.dialogue_reference[0],
+                    Dialogue.UNASSIGNED_DIALOGUE_REFERENCE,
+                ),
+                self.counterparty_from_message(message),
                 self.agent_address,
             )
             result = self_initiated_dialogue_label in self.dialogue_labels
         else:
             other_initiated_dialogue_label = DialogueLabel(
-                message.dialogue_reference, message.counterparty, message.counterparty
+                message.dialogue_reference,
+                self.counterparty_from_message(message),
+                self.counterparty_from_message(message),
```

Comment:
I'd assign to helper var and do a single call to `self.counterparty_from_message(message)`

Reply:
Done